Add momentum phase will begin to WebWheelEvent

nmoucht opened the pull request:
#### b31973ed822b67cb44fce5b2c251a895c05c3db1
<pre>
Add momentum phase will begin to WebWheelEvent
<a href="https://bugs.webkit.org/show_bug.cgi?id=295945">https://bugs.webkit.org/show_bug.cgi?id=295945</a>
<a href="https://rdar.apple.com/155832880">rdar://155832880</a>

Reviewed by Simon Fraser.

To properly implement scrollend events, we need to know if a
phase ended wheel event will result in subsequent momentum events.
Although this has not been adopted by AppKit, we can look at the
underlying IOHIDEvent. This patch adds a WillBegin phase, which will
be consulted in an upcoming patch implemnting the scrollend event for
momentum scrolls.

* Source/WebCore/PAL/pal/spi/mac/IOKitSPIMac.h:
* Source/WebCore/page/WheelEventDeltaFilter.cpp:
(WebCore::BasicWheelEventDeltaFilter::updateFromEvent):
* Source/WebCore/page/mac/WheelEventDeltaFilterMac.mm:
(WebCore::WheelEventDeltaFilterMac::updateFromEvent):
* Source/WebCore/page/scrolling/ScrollingTreeGestureState.cpp:
(WebCore::ScrollingTreeGestureState::nodeDidHandleEvent):
* Source/WebCore/platform/PlatformWheelEvent.cpp:
(WebCore::operator&lt;&lt;):
* Source/WebCore/platform/PlatformWheelEvent.h:
(WebCore::PlatformWheelEvent::isMomentumEvent const):
* Source/WebCore/platform/PlatformWheelEvent.serialization.in:
* Source/WebCore/platform/mac/ScrollingEffectsController.mm:
(WebCore::ScrollingEffectsController::handleWheelEvent):
* Source/WebKit/Shared/WebEvent.serialization.in:
* Source/WebKit/Shared/WebWheelEvent.h:
(WebKit::WebWheelEvent::isMomentumEvent const):
* Source/WebKit/Shared/mac/WebEventFactory.mm:
(WebKit::WebEventFactory::createWebWheelEvent):
* Source/WebKit/WebProcess/Plugins/PDF/PDFPlugin.mm:
(WebKit::PDFPlugin::handleWheelEvent):
* Source/WebKit/WebProcess/WebPage/MomentumEventDispatcher.cpp:
(WebKit::MomentumEventDispatcher::handleWheelEvent):
* Source/WebKitLegacy/mac/DOM/WebDOMOperations.mm:
(toNSEventPhase):

Canonical link: <a href="https://commits.webkit.org/297432@main">https://commits.webkit.org/297432@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/216296e66bd801e11005004897be2b88db235210

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/111721 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/31387 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/21862 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/117753 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/61979 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/113683 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/32075 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/39972 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/84885 "Passed tests") | [❌ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/35596 "") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/114668 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/25612 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/100551 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/65323 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/24947 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/18683 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/61585 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/94999 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/18755 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/121004 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/38767 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/28819 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/93779 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/39147 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/96806 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/93601 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/38750 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/16540 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/34797 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/18006 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/38661 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/44152 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/38315 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/41643 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/40022 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

smfr commented on Source/WebKit/Shared/mac/WebEventFactory.mm:
```diff
@@ -190,6 +190,11 @@ static int typeForEvent(NSEvent *event)
 
+        if (IOHIDEventGetScrollMomentum(ioHIDEvent.get()) & kIOHIDEventScrollMomentumWillBegin) {
+            ASSERT(momentumPhase == WebWheelEvent::Phase::PhaseNone && phase == WebWheelEvent::Phase::PhaseEnded);
+            momentumPhase = WebWheelEvent::Phase::PhaseWillBegin;
```
I'm a bit worried that there's code that checks for event. momentumPhase != None and will get confused with this change. It's new to have a non-None momentum phase on any non-momentum event.

nmoucht replied:
Added a new `isMomentumEvent` function to the events and adopted it at call sites checking momentumPhase != None.

nmoucht committed:
Add momentum phase will begin to WebWheelEvent
https://bugs.webkit.org/show_bug.cgi?id=295945
rdar://155832880

Reviewed by Simon Fraser.

To properly implement scrollend events, we need to know if a
phase ended wheel event will result in subsequent momentum events.
Although this has not been adopted by AppKit, we can look at the
underlying IOHIDEvent. This patch adds a WillBegin phase, which will
be consulted in an upcoming patch implemnting the scrollend event for
momentum scrolls.

* Source/WebCore/PAL/pal/spi/mac/IOKitSPIMac.h:
* Source/WebCore/page/WheelEventDeltaFilter.cpp:
(WebCore::BasicWheelEventDeltaFilter::updateFromEvent):
* Source/WebCore/page/mac/WheelEventDeltaFilterMac.mm:
(WebCore::WheelEventDeltaFilterMac::updateFromEvent):
* Source/WebCore/page/scrolling/ScrollingTreeGestureState.cpp:
(WebCore::ScrollingTreeGestureState::nodeDidHandleEvent):
* Source/WebCore/platform/PlatformWheelEvent.cpp:
(WebCore::operator<<):
* Source/WebCore/platform/PlatformWheelEvent.h:
(WebCore::PlatformWheelEvent::isMomentumEvent const):
* Source/WebCore/platform/PlatformWheelEvent.serialization.in:
* Source/WebCore/platform/mac/ScrollingEffectsController.mm:
(WebCore::ScrollingEffectsController::handleWheelEvent):
* Source/WebKit/Shared/WebEvent.serialization.in:
* Source/WebKit/Shared/WebWheelEvent.h:
(WebKit::WebWheelEvent::isMomentumEvent const):
* Source/WebKit/Shared/mac/WebEventFactory.mm:
(WebKit::WebEventFactory::createWebWheelEvent):
* Source/WebKit/WebProcess/Plugins/PDF/PDFPlugin.mm:
(WebKit::PDFPlugin::handleWheelEvent):
* Source/WebKit/WebProcess/WebPage/MomentumEventDispatcher.cpp:
(WebKit::MomentumEventDispatcher::handleWheelEvent):
* Source/WebKitLegacy/mac/DOM/WebDOMOperations.mm:
(toNSEventPhase):

Canonical link: https://commits.webkit.org/297432@main
